core/shared/src/main/scala/fs2/Pull.scala

```diff
@@ -33,12 +33,87 @@ import fs2.internal._
 import Pull._
 
-/** A `p: Pull[F,O,R]` reads values from one or more streams, returns a
-  * result of type `R`, and produces a `Stream[F,O]` when calling `p.stream`.
+/** A pure abstraction representing the intention to perform a sequence of actions
+  * in an effect type `F`, which emits any number of values of an output type `O`.
+  * This sequence may end up in several ways:
+  * a. it may not terminate and keep unrolling new items;
+  * b. it may be aborted by a `Throwable` error;
+  * c. be interrupted by a signal, which may include an error,
+  * d. terminate successfully and returning one single carry-over result of type `R`.
   *
-  * Any resources acquired by `p` are freed following the call to `stream`.
+  * Much like the `IO` values fromo the `cats-effect` library,
+  * `Pull` values are pure, immutable values. Thus, they preserve referential
+  * transparency and are usable in functional programming.
+  * A `Pull` is a data structure that represents just a description
+  * of a sequence of potentially side effectful computation.
   *
-  * Laws:
+  * A pull emits the output values not one-at-a-time, but one-chunk at a time.
+  * A pull only emits non-empty chunks.
+  *
+  * === Pulls variations
+  *
+  * There are some special types of Pull, which are reflected on the arguments of its type parameters:
+  *
+  * A '''pure''' pull is one that performs no effectful actions.
+  * The type of a pure pull has the `F` effect type  bound to `Nothing`.
+  * _Note_ that type `Nothing` is polykinded type, so it can also be
+  * applied as an argument to the type parameter `F[_]`.
+  * Nevertheless, for the sake of clarity, in `fs2` we use the type alias
+  * `type Pure[A] = Nothing` to represent a pure pull as `Pull[Pure, O, R]`.
+  *
+  * A '''silent''' pull is one known to emit no outputs. In a silent pull, the
+  * output type parameter `O` is bound to `Nothing`, the bottom type.
+  *
+  * A '''broken''' pull is known never to terminate successfully, which may happen
+  * because the pull is known to fail, or is known to be interrupted.
+  * {{{
+  * def infinite: Pull[Pure, Nothing, Nothing] = F.unit >> infinite
+  * }}}
+  * A broken pull may also be one that contains an action in the effect F
+  * that fails or is known not to terminate, such as `IO.never` from `cats.effect`.
+  *
+  * A '''closed''' pull is one that, if may terminate successfully with no
+  * carry-over information. The type of a closed pull has the type parameter `R`
+  * bound to `Unit`. A `Stream[F, O]` is just a wrapper  around a closed pull.
+  *
+  * === Evaluation and Compilation ===
+  *
+  * Like functional effect types (e.g. `IO` in Cats Effect),   a pull is just
+  * a passive data structure that describes a sequence of action.
+  * It is neither a running process nor a container to be eventually filled
+  * with the  result of a concurrent process (like `scala.concurrent.Future`).
+  *
+  * To "run" or "interpret" a pull, it has to be compiled.
```

Comment:
From an API perspective, pulls don't get compiled, only streams do. This doc feels more like internal documentation as to how the pull interpreter works. Not sure if we need this stuff here or if it'd be better to just talk about converting pulls to streams and then compiling. Stuff like how pulls are implemented via a monadic fold are irrelevant to users IMO.